lib/src/clixon_xml_map.c

```diff
@@ -1588,6 +1588,226 @@ yang_enum2int(yang_stmt *ytype,
+    int      i = 0;
+    int      byte = 0;
+    char   **vec = NULL;
+    char    *v;
+    char    *buffer = NULL;
+    int      nvec;
+    int      ret = 0;
+    uint32_t bitpos;
+
+    *snmplen = 0;
+
+    if ((buffer = calloc(CLIXON_BITS_POS_MAX / 8, sizeof(unsigned char))) == NULL){
+        clixon_err(OE_UNIX, errno, "calloc");
+        goto done;
+    }
+
+    if ((vec = clicon_strsep(bitsstr, " ", &nvec)) == NULL){
+        clixon_err(OE_UNIX, EINVAL, "split string failed");
+        goto done;
+    }
+
+    // Go over all set flags in given bitstring
+    for (i=0; i<nvec; i++){
+        v = clixon_trim(vec[i]);
+        if (strlen(v) > 0) {
+            if ((ret = yang_bits_pos(ytype, v, &bitpos)) < 0)
+                goto done;
+            if (ret == 0)
+                goto fail;
+
+            // Set bit at correct byte and bit position
+            byte = bitpos / 8;
+            buffer[byte] = buffer[byte] | (1 << (7 - (bitpos % 8)));
+            *snmplen = byte + 1;
+
+            if (*snmplen >= CLIXON_BITS_POS_MAX) {
+                clixon_err(OE_UNIX, EINVAL, "bit position %zu out of range. (max. allowed %d)",
+                        *snmplen, CLIXON_BITS_POS_MAX);
+                goto done;
+            }
+        }
+    }
+
+    if ((*snmpval = malloc(*snmplen)) == NULL){
+        clixon_err(OE_UNIX, errno, "calloc");
+        goto done;
+    }
+
+    memcpy(*snmpval, buffer, *snmplen);
+    retval = 1;
+
+ done:
+    if (buffer)
+        free(buffer);
+    if (vec)
+        free(vec);
+    return retval;
+
+ fail:
+    retval = 0;
+    goto done;
+}
+
+/*! Given a YANG (bits) type node and SNMP value, return the string value for all bits (flags) that are set.
+ *
+ * @param[in]  ytype   YANG type noden
+ * @param[in]  snmpval SNMP value
+ * @param[in]  snmplen length of snmpval
+ * @param[out] cb      space separated string with bit labes for all bits that are set in snmpval
+ * @retval     1       OK, result in cb
+ * @retval     0       Invalid, not found
+ * @retval    -1       Error
+ * @see yang_bitsstr2val
+ */
+int 
+yang_val2bitsstr(yang_stmt    *ytype,
+                unsigned char *snmpval,
+                size_t         snmplen,
+                cbuf          *cb)
+{
+    int        retval = -1;
+    int        is_first = 1;
+    int        ret = 0;
+    int        byte = 0;
+    char      *reason;
+    yang_stmt *yprev = NULL;
+    yang_stmt *ypos;
+    uint32_t   bitpos = 0;
+
+    if (cb == NULL){
+        clixon_err(OE_UNIX, EINVAL, "cb is NULL");
+        goto done;
+    }
+
+    // Go over all defined bits and check if it is seet in intval
+    while ((yprev = yn_each(ytype, yprev)) != NULL && byte < snmplen){
+        if (yang_keyword_get(yprev) == Y_BIT) {
+            // Use position from Y_POSITION statement if defined
+            if ((ypos = yang_find(yprev, Y_POSITION, NULL)) != NULL){
+                if ((ret = parse_uint32(yang_argument_get(ypos), &bitpos, &reason)) < 0){
```

Comment:
reason is not free:d.
Preferable initialize it to NULL and free it in the done: claues